src/jdk.unsupported/share/classes/sun/misc/Unsafe.java

```diff
@@ -1255,12 +1731,207 @@ public void fullFence() {
      * @throws IllegalArgumentException if {@code directBuffer} is non-direct,
      * or is a {@link java.nio.Buffer#slice slice}, or is a
      * {@link java.nio.Buffer#duplicate duplicate}
+     *
+     * @deprecated Use a {@link MemorySegment} allocated in an {@link Arena} with the
+     * appropriate temporal bounds. The {@link MemorySegment#asByteBuffer()} method
+     * wraps a memory segment as a {@code ByteBuffer} to allow interop with existing
+     * code.
+     *
      * @since 9
      */
+    @Deprecated(since="23", forRemoval=true)
     public void invokeCleaner(java.nio.ByteBuffer directBuffer) {
         if (!directBuffer.isDirect())
-            throw new IllegalArgumentException("buffer is non-direct");
-
+            throw new IllegalArgumentException("Not a direct buffer");
+        beforeMemoryAccess();
         theInternalUnsafe.invokeCleaner(directBuffer);
     }
+
+    // Instructure for --sun-misc-unsafe-memory-access=<value> command line option.
+
+    private static final Object MEMORY_ACCESS_WARNED_BASE;
+    private static final long MEMORY_ACCESS_WARNED_OFFSET;
+    static {
+        try {
+            Field field = Unsafe.class.getDeclaredField("memoryAccessWarned");
+            MEMORY_ACCESS_WARNED_BASE = theInternalUnsafe.staticFieldBase(field);
+            MEMORY_ACCESS_WARNED_OFFSET = theInternalUnsafe.staticFieldOffset(field);
+        } catch (Exception e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+    // set to true by first usage of memory-access method
+    private static @Stable boolean memoryAccessWarned;
```

Comment:
Very nice code. The fact that you only need to report first use (and not first use per module, like in FFM) simplify things a bit (and, most importantly, avoids the need to make all unsafe methods caller sensitive!).